### ISSUE TITLE
WARNING: attempting to recover from OOM in forward/backward pass

### ISSUE THREAD

**Issue by pzhren** [krasserm/fairseq-image-captioning#23]
Hi, I encountered some errors during the Self-critical sequence training stage: 
WARNING: attempting to recover from OOM in forward/backward pass
Is this because the GPU memory is not enough? It feels very strange, because sometimes it is normal.

**Comment by krasserm** [comment]
> Is this because the GPU memory is not enough?

Yes, this is the reason. The settings documented in the README are appropriate for 2 GTX 1080 cards (8 GB each).

**Comment by pzhren** [comment]
In fact, I used 3 GPUs, each of which is 11g. The strange thing is that sometimes it works normally, and sometimes it is reported that the storage is insufficient.

**Comment by krasserm** [comment]
Did you pre-train the model with [CE loss](https://github.com/krasserm/fairseq-image-captioning/blob/master/README.md#cross-entropy-loss) before running SCST?

**Comment by pzhren** [comment]
Yes. I passed --max-sentences 2, and it ran normally, but I was worried that it would affect performance. I don't know if it will have a significant impact? Besides, why not use .checkpoint/checkpoint_best.pt, is this not the best weight?

**Comment by krasserm** [comment]
Convergence improves with higher --max-sentences values (but also requires more memory). A value of 5 should work fine on 11 GB cards. 

Regarding `checkpoint_best.pt`, this is the checkpoint with the best CE validation loss, but not necessarily with the best CIDEr score (or any other evaluation metric). Checkpoint selection based on a user defined metric should be automated but I had other priorities in the past. Hope I can resume work on it anytime soon. Pull requests are welcome too, of course!

**Comment by pzhren** [comment]
I see, thank you.

**Comment by pzhren** [comment]
![image](https://user-images.githubusercontent.com/34993251/98348684-09483180-2054-11eb-9400-1176b431dedb.png)
In fact, we found that when SCST was running, one of the GPU memory loads suddenly became too high. There is a serious load imbalance between GPUs. Do you have a good solution?

**Comment by pzhren** [comment]
Don't worry, I found that during the running process, the memory usage gradually increased. This is the running state at --max-sentence 3.
![image](https://user-images.githubusercontent.com/34993251/98351245-68f40c00-2057-11eb-8303-4147c6651b65.png)

**Comment by krasserm** [comment]
What is the frequency of OOMs when you run with --max-sentences 5 or 8?

**Comment by pzhren** [comment]
Almost every time I encounter it, the strange thing is that it reports a memory error after SCST runs one or two.
